### PR TITLE
[Access] Index collections from execution data

### PR DESCRIPTION
**Pull request by peterargue**
Closes: https://github.com/onflow/flow-go/issues/4829

### CODE REVIEW

**Review by zhangchiqing — APPROVED** [review]

File: engine/access/ingestion/engine.go [hunk 1]
```diff
@@ -676,113 +696,127 @@ func (e *Engine) requestMissingCollections(ctx context.Context) error {
...
 	if err != nil {
-		if !errors.Is(err, storage.ErrNotFound) {
-			logError(err)
-			return
+		return fmt.Errorf("failed to find last full block received height: %w", err)
+	}
+
+	// if more contiguous blocks are now complete, update db
+	if newLastFullHeight > lastFullHeight {
+		err = e.blocks.UpdateLastFullBlockHeight(newLastFullHeight)
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/access/ingestion/engine.go:721]: good call out. I opened an issue to track this: https://github.com/onflow/flow-go/issues/5185

File: engine/access/ingestion/engine.go [hunk 2]
```diff
@@ -676,113 +696,127 @@ func (e *Engine) requestMissingCollections(ctx context.Context) error {
...
-	// additionally, if more than threshold blocks have missing collection OR collections are missing since defaultMissingCollsForAgeThreshold, re-request those collections
-	if incompleteBlksCnt >= defaultMissingCollsForBlkThreshold || (finalizedHeight-lastFullHeight) > uint64(defaultMissingCollsForAgeThreshold) {
+	// additionally, if more than threshold blocks have missing collections OR collections are
+	// missing since defaultMissingCollsForAgeThreshold, re-request those collections
+	if incompleteBlksCnt >= defaultMissingCollsForBlkThreshold ||
+		(finalizedHeight-lastFullHeight) > defaultMissingCollsForAgeThreshold {
 		// warn log since this should generally not happen
 		e.log.Warn().
+			Uint64("last_full_blk_height", lastFullHeight).
 			Int("missing_collection_blk_count", incompleteBlksCnt).
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/access/ingestion/engine.go:801]: [d7052d2](https://github.com/onflow/flow-go/pull/5163/commits/d7052d2945436a5552b67bf5bc7e04bf94f060c3)

File: module/state_synchronization/indexer/indexer_core.go [hunk 3]
```diff
@@ -159,6 +161,27 @@ func (c *IndexerCore) IndexBlockData(data *execution_data.BlockExecutionDataEnti
...
+	g.Go(func() error {
+		start := time.Now()
+
+		// index all collections except the system chunk
+		// Note: the access ingestion engine also indexes collections, starting when the block is
+		// finalized. This process can fall behind due to the node being offline, resource issues
+		// or network congestion. This indexer ensures that collections are never farther behind
+		// than the latest indexed block. Calling the collection handler with a collection that
+		// has already been indexed is a noop.
+		for _, chunk := range data.ChunkExecutionDatas[0 : len(data.ChunkExecutionDatas)-1] {
```
[the remark this answers is not in the document]
Comment (peterargue) [reply module/state_synchronization/indexer/indexer_core.go]: [d7052d2](https://github.com/onflow/flow-go/pull/5163/commits/d7052d2945436a5552b67bf5bc7e04bf94f060c3)

File: cmd/access/node_builder/access_node_builder.go [hunk 4]
```diff
@@ -716,14 +716,14 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 
 					checkpointHeight := builder.SealedRootBlock.Header.Height
 
-					buutstrap, err := pStorage.NewRegisterBootstrap(pdb, checkpointFile, checkpointHeight, builder.Logger)
+					bootstrap, err := pStorage.NewRegisterBootstrap(pdb, checkpointFile, checkpointHeight, builder.Logger)
```
Comment (sideninja) [inline cmd/access/node_builder/access_node_builder.go:719]: that was an itch for me 😄

File: engine/access/ingestion/engine.go [hunk 5]
```diff
@@ -48,9 +51,10 @@ const defaultQueueCapacity = 10_000
 
 var defaultCollectionCatchupTimeout = collectionCatchupTimeout
 var defaultCollectionCatchupDBPollInterval = collectionCatchupDBPollInterval
-var defaultFullBlockUpdateInterval = fullBlockUpdateInterval
+var defaultFullBlockRefreshInterval = fullBlockRefreshInterval
+var defaultMissingCollsRequestInterval = missingCollsRequestInterval
 var defaultMissingCollsForBlkThreshold = missingCollsForBlkThreshold
-var defaultMissingCollsForAgeThreshold = missingCollsForAgeThreshold
+var defaultMissingCollsForAgeThreshold uint64 = missingCollsForAgeThreshold
```
Comment (sideninja) [inline engine/access/ingestion/engine.go:57]: maybe:
```
var (
 ...
)
```

File: engine/access/ingestion/engine.go [hunk 6]
```diff
@@ -676,113 +696,128 @@ func (e *Engine) requestMissingCollections(ctx context.Context) error {
...
+			Uint64("last_full_blk_height", lastFullHeight).
 			Int("missing_collection_blk_count", incompleteBlksCnt).
-			Int("threshold", defaultMissingCollsForBlkThreshold).
-			Uint64("last_full_blk_height", latestFullHeight).
+			Int("missing_collection_count", len(allMissingColls)).
 			Msg("re-requesting missing collections")
 		e.requestCollectionsInFinalizedBlock(allMissingColls)
 	}
 
-	e.log.Debug().Uint64("last_full_blk_height", latestFullHeight).Msg("updated LastFullBlockReceived index")
```
Comment (sideninja) [inline engine/access/ingestion/engine.go:771]: no need for this anymore?

**Review by sideninja — APPROVED** [review]

File: engine/access/ingestion/engine.go (hunk 5 above)
Comment (peterargue) [reply engine/access/ingestion/engine.go]: [953ac9e](https://github.com/onflow/flow-go/pull/5163/commits/953ac9ede3b3c9ecf6db19eebc3d52082e29ffd1)

**Pull request merged**